File: src/main/java/me/redth/statsoverlay/config/OverlayHUD.java
package me.redth.statsoverlay.config;

import cc.polyfrost.oneconfig.hud.BasicHud;
import cc.polyfrost.oneconfig.libs.universal.UGraphics;
import cc.polyfrost.oneconfig.libs.universal.UMatrixStack;
import cc.polyfrost.oneconfig.platform.Platform;
import cc.polyfrost.oneconfig.utils.hypixel.HypixelUtils;
import me.redth.statsoverlay.StatsOverlay;
import me.redth.statsoverlay.stat.DataList;
import me.redth.statsoverlay.stat.PlayerData;
import me.redth.statsoverlay.stat.StatType;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class OverlayHUD extends BasicHud {
    private final List<PlayerData> display = new ArrayList<>();
    private final Map<StatType, Integer> widthMap = new HashMap<>();

    public OverlayHUD() {
        super(true);
    }

    @Override
    protected void draw(UMatrixStack matrices, float x, float y, float scale, boolean example) {
        UGraphics.GL.pushMatrix();
        UGraphics.GL.translate(x, y, 0);
        UGraphics.GL.scale(scale, scale, 1);

        int x1 = 0;

        for (StatType type : StatType.values()) {
            if (!type.shouldShow()) continue;
            int width = widthMap.get(type);

            drawText(type.name, x1 + 2, 0, 0xFFFFFFFF);

            int y1 = 10;
            for (PlayerData stats : display) {
                if (type.rightAlign)
                    drawRightedText(stats.getStatFormatted(type), x1 + width - 2, y1, stats.getStatColor(type));
                else
                    drawText(stats.getStatFormatted(type), x1 + 2, y1, stats.getStatColor(type));
                y1 += 10;
            }
            x1 += width;
        }

        UGraphics.GL.popMatrix();
    }

    @Override
    protected float getWidth(float scale, boolean example) {
        if (widthMap == null) return 0F;

        if (DataList.altered) {
            update();
            DataList.altered = false;
        }

        float cacheWidth = 0;

        for (StatType type : StatType.values()) {
            if (!type.shouldShow()) continue;
            cacheWidth += widthMap.get(type);
        }

        return cacheWidth * scale;
    }

    @Override
    protected float getHeight(float scale, boolean example) {
        if (display == null) return 0F;

        return (display.size() * 10 + 8) * scale;
    }

    private void update() {
        DataList.update(display);

        for (StatType type : StatType.values()) {
            widthMap.put(type, calculateWidth(type));
        }
    }

    private int calculateWidth(StatType type) {
        int width = Platform.getGLPlatform().getStringWidth(type.name);
        for (PlayerData stats : display) {
            width = Math.max(width, Platform.getGLPlatform().getStringWidth(stats.getStatFormatted(type)));
        }
        return width + 4;
    }

    @Override
    protected boolean shouldShow() {
        return super.shouldShow() && (OverlayConfig.showAnyWhere || (HypixelUtils.INSTANCE.isHypixel() && StatsOverlay.inBW)) && OverlayConfig.isKeyPressed();
    }

    private void drawCenteredText(String text, float x, float y, int color) {
        int width = Platform.getGLPlatform().getStringWidth(text);
        Platform.getGLPlatform().drawText(text, x - width / 2.0F, y, color, true);
    }

    private void drawRightedText(String text, float x, float y, int color) {
        int width = Platform.getGLPlatform().getStringWidth(text);
        Platform.getGLPlatform().drawText(text, x - width, y, color, true);
    }

    private void drawText(String text, float x, float y, int color) {
        Platform.getGLPlatform().drawText(text, x, y, color, true);
    }
}

File: src/main/kotlin/me/redth/statsoverlay/config/OverlayHUD.kt
package me.redth.statsoverlay.config

import cc.polyfrost.oneconfig.events.EventManager
import cc.polyfrost.oneconfig.events.event.Stage
import cc.polyfrost.oneconfig.events.event.TickEvent
import cc.polyfrost.oneconfig.hud.BasicHud
import cc.polyfrost.oneconfig.libs.eventbus.Subscribe
import cc.polyfrost.oneconfig.libs.universal.UMatrixStack
import cc.polyfrost.oneconfig.utils.dsl.mc
import me.redth.statsoverlay.EventListener
import me.redth.statsoverlay.data.Alignment
import me.redth.statsoverlay.data.CompiledGrid
import me.redth.statsoverlay.data.ColoredText
import me.redth.statsoverlay.data.FetchQueue
import net.minecraft.client.renderer.GlStateManager

private const val ROW_HEIGHT = 10

class OverlayHUD : BasicHud(true) {
    @Transient
    private var compiledGrid: CompiledGrid? = null

    init {
        EventManager.INSTANCE.register(this)
    }

    @Subscribe
    fun onTick(event: TickEvent) {
        if (event.stage != Stage.END) return
        compiledGrid = FetchQueue.makeGrid()
    }

    override fun draw(matrices: UMatrixStack, x: Float, y: Float, scale: Float, example: Boolean) {
        val grid = compiledGrid ?: return

        GlStateManager.pushMatrix()
        GlStateManager.translate(x, y, 0f)
        GlStateManager.scale(scale, scale, 1f)

        var yNow = 0
        for (row in grid.grid) {
            var xNow = 0
            for ((index, text) in row.withIndex()) {
                val width = grid.widthMap[index]
                drawText(text, xNow, yNow, width)
                xNow += width
            }
            yNow += ROW_HEIGHT
        }


        GlStateManager.popMatrix()
    }

    override fun getWidth(scale: Float, example: Boolean) = (compiledGrid?.width?.toFloat() ?: 0f) * scale
    override fun getHeight(scale: Float, example: Boolean) = (compiledGrid?.height?.toFloat() ?: 0f) * scale

    override fun shouldShow() = super.shouldShow() && EventListener.shouldShow()

    private fun drawText(text: ColoredText, x: Int, y: Int, width: Int) {
        val offset = when (text.alignment) {
            Alignment.LEFT -> 0f
            Alignment.CENTER -> width / 2f - text.width / 2f
            Alignment.RIGHT -> width.toFloat() - text.width.toFloat()
        }
        mc.fontRendererObj.drawString(text.text, x.toFloat() + offset, y.toFloat(), text.color, true)
    }

}
